### ISSUE TITLE
Retrieve number of subscribers

### ISSUE THREAD

**Issue by sublee** [zeromq/libzmq#2095]
I suggest a function which retrieves the number of subscribers on `PUB` or `XPUB` sockets.

I want to reduce serialization cost if a message wouldn't be published because any subscribers haven't subscribed the prefix.  To achieve it, I read subscription and unsubscription events from `XPUB` sockets and manage them with a Trie by my hand.

But I know there're already Tries in `PUB` of `XPUB` sockets to manage `SUB` or `XSUB` sockets' subscriptions.  How about I can retrieve the number of subscribers on `PUB` or `XPUB` sockets?  If the number is Zero, I can drop meaningless serializations immediately.

Here's an example:

```python
if zmq.subscribers(pub, prefix) != 0:
    pub.send_multipart([prefix, serialize_expansively(msg)])
```

**Comment by bluca** [comment]
Haven't looked in the trie code in a while, but off the top of my head I can think immediately of 2 issues:

1) If the trie is managed in the I/O thread, there are immediately concurrent access problems to keep in mind
2) Race condition could be an issue since locking is out of the question, so for example you could have subscribers leaving after you check but before you send

That said, if you have an implementation in mind through zmq_getsockopt that doesn't affect performances and it's concurrency-safe please do send a PR and we'll merge it

**Comment by somdoron** [comment]
Trie is on the user thread, so you might add a socket option for that.
Anyway I'm not a big fun of this. Is it really your bottleneck now?

**Comment by sublee** [comment]
Currently, I'm making a game server.  Serialization is a common bottleneck in game servers.  For example, to serialize game players or inventories containing 1k or more items.  So I think it might can be useful.

**Comment by hnwyllmm** [comment]
you can use a monitor socket to get the number of PUB clients which let you know new connection and client disconnection.

**Comment by sublee** [comment]
@hnwyllmm Can I get the number of subscribers for specific prefix through a monitor socket?

**Comment by hnwyllmm** [comment]
you only can count the connections number by monitor socket.
By monitor socket, you can catch the event of new connection or disconnection and then count the connection number which is the number of subscribers.

**Comment by sublee** [comment]
@hnwyllmm Yes, you are right.  But that is not what I need.  I should detect the number of subscribers by specific prefix instead of just all of the peer connections.

**Comment by hnwyllmm** [comment]
sorry, I misunderstand you.
Does the 'prefix' is the client's host ip address?
If that true, you can modify the tcp_listener.cpp:122(libzmq master) to send the peer ip address to monitor socket in the new connection event and the disconnection event in called in stream_engine_t::error.

The code location are below:
connection event:
[https://github.com/zeromq/libzmq/blob/master/src/tcp_listener.cpp#122](https://github.com/zeromq/libzmq/blob/master/src/tcp_listener.cpp#122)
disconnection event:
[https://github.com/zeromq/libzmq/blob/master/src/stream_engine.cpp#972](https://github.com/zeromq/libzmq/blob/master/src/stream_engine.cpp#972)

**Comment by sublee** [comment]
@hnwyllmm No, it doesn't.  The `prefix` is the first _n_ bytes of the messages which I will send through a `PUB` socket.  A `SUB` socket subscribes messages by their prefixes.  For example, if a `SUB` socket subscribes `a`, it would receive `apple` or `alice`, but it wouldn't receives `bob` or `bless` and so on, unless it starts to subscribe `b`.

**Comment by hnwyllmm** [comment]
Eh, Is the prefix the subscribers settings? If so, the count method is in zmq::xpub_t::xsend and you can create a new method to retrieve the number, or a new option in getsockopt.
code:
[https://github.com/zeromq/libzmq/blob/master/src/xpub.cpp#229](https://github.com/zeromq/libzmq/blob/master/src/xpub.cpp#229)

**Comment by bluca** [comment]
@sublee have you looked into ZMQ_XPUB_VERBOSE ?

https://github.com/zeromq/libzmq/blob/master/doc/zmq_setsockopt.txt#L916

If you switch to XPUB and set that should let you receive the subscribe messages when you read on the pub socket

**Comment by somdoron** [comment]
Maybe we can implement an hasOut variant that accept a parameter (or two
parameters, one for the size).
It can also be useful for Router and Server sockets.

we can use a new socket option for this.
